File: ruby/test/test_facade.rb
# frozen_string_literal: true

require "minitest/autorun"
require "baseh"

# Zero-config facade over the frozen baseh-expandable-v1 profile: Baseh.encode
# and Baseh.decode share one lazily built codec instance.
class TestFacade < Minitest::Test
  def expandable
    @expandable ||= Baseh::Baseh.new(Baseh.baseh_expandable_v1)
  end

  def assert_baseh_error(code, &block)
    error = assert_raises(Baseh::BasehError, &block)
    assert_equal code, error.code
  end

  def test_encode_returns_a_string
    assert_instance_of String, Baseh.encode(0)
    assert_instance_of String, Baseh.encode(123_456)
  end

  def test_decode_encode_round_trip
    [0, 1, 42, 813, 1_156, 123_456, 1_000_000_000_000].each do |id|
      assert_equal id, Baseh.decode(Baseh.encode(id)).id
    end
  end

  def test_decode_returns_the_instance_decode_result
    result = Baseh.decode(Baseh.encode(123_456))
    assert_instance_of Baseh::Baseh::DecodeResult, result
    assert_equal 123_456, result.id
    assert_equal Baseh.encode(123_456), result.canonical_code
    assert_equal false, result.corrected
  end

  def test_facade_agrees_with_a_manual_default_profile_instance
    [0, 1, 123_456, 1_000_000_000_000].each do |id|
      assert_equal expandable.encode(id: id), Baseh.encode(id)
      assert_equal expandable.encode(id: id), Baseh.decode(Baseh.encode(id)).canonical_code
    end
  end

  def test_errors_surface_like_the_instance_api
    assert_baseh_error("OUT_OF_RANGE") { Baseh.encode(-1) }
    assert_baseh_error("INVALID_LENGTH") { Baseh.decode("AB") }
    assert_baseh_error("INVALID_CHARACTER") { Baseh.decode("!!!!") }
    assert_baseh_error("INVALID_CHECKSUM") { Baseh.decode("ZZZZ") }
  end

  def test_decode_passes_keyword_options_through
    code = Baseh.encode(123_456)
    assert_equal 123_456, Baseh.decode(" #{code} ", accept_spaces: true).id
  end
end

File: python/tests/test_facade.py
"""Facade tests: the zero-config package-level encode/decode agree with a
manually-constructed Baseh on the frozen expandable v1 tier, round trip ids
including 0 and large values, and surface errors like the instance API."""

import unittest

from baseh import (
    Baseh,
    BasehError,
    DecodeResult,
    INVALID_CHECKSUM,
    OUT_OF_RANGE,
    baseh_expandable_v1,
    decode,
    encode,
    validate,
)


class TestFacade(unittest.TestCase):
    def test_encode_returns_str(self):
        self.assertIsInstance(encode(42), str)

    def test_round_trip(self):
        for id in (0, 1, 42, 39303, 39304, 1_000_000, 1_336_335, 2**40):
            result = decode(encode(id))
            self.assertIsInstance(result, DecodeResult)
            self.assertEqual(result.id, id)

    def test_agrees_with_manual_instance(self):
        codec = Baseh(baseh_expandable_v1())
        for id in (0, 7, 12345, 999_999_999):
            code = encode(id)
            self.assertEqual(code, codec.encode(id))
            self.assertEqual(decode(code), codec.decode(code))

    def test_decode_error_matches_instance_api(self):
        codec = Baseh(baseh_expandable_v1())
        code = encode(5)
        bad = code[:-1] + ("0" if code[-1] != "0" else "1")
        with self.assertRaises(BasehError) as facade_ctx:
            decode(bad)
        with self.assertRaises(BasehError) as instance_ctx:
            codec.decode(bad)
        self.assertEqual(facade_ctx.exception.code, instance_ctx.exception.code)
        self.assertEqual(facade_ctx.exception.code, INVALID_CHECKSUM)

    def test_encode_out_of_range(self):
        with self.assertRaises(BasehError) as ctx:
            encode(-1)
        self.assertEqual(ctx.exception.code, OUT_OF_RANGE)

    def test_validate_accepts_valid_code(self):
        code = encode(42)
        self.assertEqual(validate(code), {"valid": True, "canonical_code": code})

    def test_validate_rejects_invalid_code_without_raising(self):
        code = encode(5)
        bad = code[:-1] + ("0" if code[-1] != "0" else "1")
        result = validate(bad)
        self.assertEqual(result["valid"], False)
        self.assertEqual(result["reason"], INVALID_CHECKSUM)

    def test_validate_agrees_with_manual_instance(self):
        codec = Baseh(baseh_expandable_v1())
        code = encode(99)
        for candidate in (code, code.lower()):
            self.assertEqual(validate(candidate), codec.validate(candidate))


if __name__ == "__main__":
    unittest.main()
